Mathlib/FieldTheory/JacobsonNoether.lean

```diff
@@ -0,0 +1,233 @@
+      simp only [hm2 (p^ m + 1) (by linarith), LinearMap.zero_apply]
+    classical
+    refine ⟨Nat.find exist, ⟨(Nat.find_spec exist).1, ?_, (Nat.find_spec exist).2⟩⟩
+    set t := (Nat.find exist - 1 : ℕ) with ht
+    by_cases choice : 0 < t
+    · have := @Nat.find_min (H := exist) _ t ?_
+      · exact (@Nat.sub_add_cancel (Nat.find exist) 1 (by omega) ▸ ht ▸ not_and.1 this) choice
+      · exact Nat.sub_one_lt <| ne_of_gt (Nat.find_spec exist).1
+    · rw [not_lt, Nat.le_zero] at choice
+      have := Nat.eq_add_of_sub_eq (Nat.find_spec exist).1 ht.symm
+      simp only [gt_iff_lt, choice, Nat.succ_eq_add_one, zero_add] at this
+      rw [this]
+      simp only [Function.iterate_one, ne_eq]
+      exact hb1
+  -- We define `c` to be the value that we proved above to be non-zero.
+  set c := ((δ a) ^ n) b with hc_def
+  letI : Invertible c := ⟨c⁻¹, inv_mul_cancel₀ (hb.1), mul_inv_cancel₀ (hb.1)⟩
+  have hc : c * a = a * c := by
+    symm; apply eq_of_sub_eq_zero
+    rw [← f_def a c, ← g_def a c, ← δ_def, δ_iterate_succ a b n, hb.2]
+  -- We now make some computation to obtain the *absurd* equation `final_eq`, contradiction.
+  set d := c⁻¹ * a * (δ a) ^[n-1] b with hd_def
+  have hc': c⁻¹ * a = a * c⁻¹ := by
+    apply_fun (c⁻¹ * · * c⁻¹) at hc
+    rw [← mul_assoc, inv_mul_cancel₀ hb.1, one_mul, mul_assoc, mul_assoc,
+      mul_inv_cancel₀ hb.1, mul_one] at hc
+    exact hc.symm
+
+  have c_eq : a * (δ a) ^[n - 1] b - (δ a) ^[n - 1] b * a = c := by
+    rw [hc_def, ← (LinearMap.pow_apply (δ a) (n - 1) b), ← Nat.sub_add_cancel hn,
+      show (δ a ^ (n - 1 + 1)) b = (δ a) ((δ a ^ (n - 1)) b) by rw [δ_iterate_succ]]
+    simp only [add_tsub_cancel_right, δ_def, f_def, g_def]
+  have eq1 : c⁻¹ * a * (δ a)^[n - 1] b - c⁻¹ * (δ a)^[n - 1] b * a = 1 := by
+    simp_rw [mul_assoc, (mul_sub_left_distrib c⁻¹ _ _).symm, c_eq, inv_mul_cancel_of_invertible]
+
+  have deq : a * d - d * a = a := by
+    calc
+      _ = a * ((c⁻¹ * a * (δ a) ^[n - 1] b) - (c⁻¹ * (δ a) ^[n - 1] b * a)) := by
+        simp_rw [hd_def, hc', mul_assoc, ← mul_sub_left_distrib]
+      _ = _ := by simp only [eq1, mul_one]
+  apply_fun (a⁻¹ * · ) at deq
+  rw [mul_sub, ← mul_assoc, inv_mul_cancel₀ ha₀, one_mul, ← mul_assoc, sub_eq_iff_eq_add] at deq
+  obtain ⟨r, hr⟩ := (exists_pow_mem_center_ofInseparable p d hinsep)
+  apply_fun fun x => x ^ (p ^ r) at deq
+  rw [add_pow_char_pow_of_commute D 1 _ (Commute.one_left _) , one_pow,
+    ← conj_nonComm_Algebra (ha := ha₀), ← hr.comm, mul_assoc, inv_mul_cancel₀ ha₀, mul_one,
+    self_eq_add_left] at deq
+  exact one_ne_zero deq
+
+variable (D) in
+/-- For a non-commutative finite dimensional division algebra D (with base ring being its center),
+  there exist a separable element xover its center-/
+theorem Jacobson_Noether (H : k ≠ (⊤ : Subring D)) :
+    ∃ x : D, x ∉ k ∧ IsSeparable k x := by
+  obtain ⟨p, hp⟩ := CharP.exists D
+  rcases CharP.char_is_prime_or_zero D p with h | h
```

Comment:
I think it's better to use `ExpChar` instead, by this way the char 0 and char p proof can probably be unified.

Reply:
Sorry I didn't catch it, why char 0 and char p can be unified through `ExpChar`? I tried to change the statement of `JacobsonNoether_charZero` and `JacobsonNoether_charP` into `ExpChar`. But when I attempted to use them in `Jacobson_Noether`, I still need `cases` to split the condition and apply two lemmas accordingly.

Reply:
I mean you don't need to separate the proof of `JacobsonNoether_charZero` and `JacobsonNoether_charP`. Instead, always use the proof of `JacobsonNoether_charP` but use `ExpChar` instead. The [isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem) is stated using `ExpChar`, hence unifies char 0 and char p case. In the char 0 case its `ExpChar` is 1, which makes your proof of `JacobsonNoether_charP` still works for char 0 case.

Reply:
> I mean you don't need to separate the proof of `JacobsonNoether_charZero` and `JacobsonNoether_charP`. Instead, always use the proof of `JacobsonNoether_charP` but use `ExpChar` instead. The [isPurelyInseparable_iff_pow_mem](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#isPurelyInseparable_iff_pow_mem) is stated using `ExpChar`, hence unifies char 0 and char p case. In the char 0 case its `ExpChar` is 1, which makes your proof of `JacobsonNoether_charP` still works for char 0 case.

Thanks! I've finished major migration in [the newest commit](https://github.com/leanprover-community/mathlib4/pull/16525/commits/7555cf451433a9626b1f78617184077df5abf847).